### PR TITLE
Start of non-WebMercatorQuad TMS support

### PR DESCRIPTION
**Pull request by AndrewAnnex**
* initial fixes for #202 #207 and #209
* adds cached to_rio_crs method
* updates get_footprints, get_dataset_info and tile_to_bounds to accept tms kwarg
* deprecates and removes assets_for_tile
* updates all code to use either assets_for_bbox or assets_for_point

Now, all I did so far to deprecate assets_for_tile was swap out usages directly for get_assets. Maybe the intent was to prevent all external direct-tile access? Looking for feedback on things I missed and especially on the usage of attr for the tms attribtue which I am not familiar with.

Still need to add tests with other tms specs, thinking making a custom epsg:4087 would be a good approach or any other built in morecantile tms.

### CODE REVIEW

File: cogeo_mosaic/backends/base.py [hunk 1]
```diff
@@ -126,19 +125,25 @@ def update(
...
         self.write(overwrite=True)
 
-    def assets_for_tile(self, x: int, y: int, z: int) -> List[str]:
-        """Retrieve assets for tile."""
-        return self.get_assets(x, y, z)
-
-    def assets_for_point(self, lng: float, lat: float) -> List[str]:
+    def assets_for_point(self, lng: float, lat: float, coord_crs: CRS = WGS84_CRS) -> List[str]:
         """Retrieve assets for point."""
+        lng, lat = transform_point(lng, lat, to_rio_crs(coord_crs), self.geographic_crs)
```
Comment (vincentsarago) [inline cogeo_mosaic/backends/base.py:130]: 🤔  I would assume `CRS` to always be in form of rasterio crs as input, so if there is a transformation to be made using the tms's CRS to rasterio's CRS.

If it helps we should add `rasterio_geographic_crs` in morecantile model.

having both pyproj and rasterio within the same module worries me!

**Comment by vincentsarago** [comment]
once https://github.com/developmentseed/cogeo-mosaic/pull/212 get merged I'll finish this PR 😄

**Comment by vincentsarago** [comment]
@AndrewAnnex I'm working on this! please do not push more change 🙏

File: cogeo_mosaic/backends/base.py [hunk 2]
```diff
@@ -57,29 +59,36 @@ class BaseBackend(BaseReader):
     ] = attr.ib(default=Reader)
     reader_options: Dict = attr.ib(factory=dict)
 
-    geographic_crs: CRS = attr.ib(default=WGS84_CRS)
-
-    # TMS is outside the init because mosaicJSON and cogeo-mosaic only
-    # works with WebMercator for now.
+    # TMS Should be the same as the Mosaic's TMS
     tms: TileMatrixSet = attr.ib(init=False, default=WEB_MERCATOR_TMS)
```
Comment (vincentsarago) [inline cogeo_mosaic/backends/base.py:63]: This should be updated (I reverted the change).

The TMS is defined by the Mosaic, but this is only supported in mosaicjson 0.0.3

**Comment by vincentsarago** [comment]
@AndrewAnnex IMO we are trying to do to many thing in one PR. I thing we should take a step back and do:
- create a PR for the `mosaic.py` code (from the PR) 
- update `mosaic.py` to make it compatible with mosaicjson 0.0.3

Once ☝️ is done we can work on the Backends.

I'm going to work on this over the weekend or on Monday

**Comment by AndrewAnnex** [comment]
@vincentsarago been stuck doing other things and catching up, what's the status of your suggestion above? sounds like you got the 2nd bullet point done?

**Comment by vincentsarago** [comment]
@AndrewAnnex 👋 
so I've done https://github.com/developmentseed/cogeo-mosaic/pull/214 and I haven't had time to work on the backend yet.

I'm going to FOSS4G next week but after I'll have time to focus on this

**Comment by AndrewAnnex** [comment]
Okay sounds good, no rush obv just wanted to check in- Dr. Andrew AnnexOn Jun 23, 2023, at 10:08 AM, Vincent Sarago ***@***.***> wrote:﻿
@AndrewAnnex 👋
so I've done #214 and I haven't had time to work on the backend yet.
I'm going to FOSS4G next week but after I'll have time to focus on this

**Comment by vincentsarago** [comment]
This was done in other PR 🙏

**Pull request closed without merging**